doc/proposals/composable-integration-in sdk.md

```diff
@@ -0,0 +1,173 @@
+
+
+### Implementation Details/Notes/Constraints (optional)
+
+The Composable SDK offers the following types.
+
+```golang
+type ObjectRef struct {
+	GetValueFrom ComposableGetValueFrom `json:"getValueFrom"`
+}
+
+type ComposableGetValueFrom struct {
+	Kind               string   `json:"kind"`
+	APIVersion         string   `json:"apiVersion,omitempty"`
+	Name               string   `json:"name,omitempty"`
+	Labels             []string `json:"labels,omitempty"`
+	Namespace          string   `json:"namespace,omitempty"`
+	Path               string   `json:"path"`
+	FormatTransformers []string `json:"format-transformers,omitempty"`
+}
+```
+
+An `ObjectRef` can be used to specify the type of any field of a CRD definition, allowing the value to be determined dynamically.
+For a detailed explanation of how to specify an object reference according to this schema, see [here](https://github.com/IBM/composable/blob/master/README.md#getvaluefrom-elements).
+
+The Composable SDK offers the following function for resolving the value of cross-resource references.
+
+```golang
+func ResolveObject(r client.Client, config *rest.Config, object interface{}, resolved interface{}, composableNamespace string) *ComposableError 
+```
+
+The function `ResolveObject` takes a Kubernetes client and configuration, and `object` to resolve, and a blank object
+`resolved` that will contain the result, as well as a namespace. The namespace is the default used when references
+do not specify one. This function will cast the result to the type of the `resolved` object, provided that
+appropriate data transformers have been included in the reference definitions (see [tutorial](https://github.com/IBM/composable/blob/master/sdk/docs/tutorial.md) for an example).
+
+The `ResolveObject` function uses caching for looking up object kinds, as well as for the objects themselves, in order
+to ensure that a consistent view of the data is obtained. If any data is not available at the time of the lookup,
+it returns an error. So this function either resolves the entire object or it doesn't -- there are no partial results.
+The cache lasts only for the duration of the `ResolveObject` call.
+
+Our proposal is to have a tight integration of `ResolveObject` into the API offered by the Operator-SDK to interact with
+Kubernetes objects.
+
+The return value is a `ComposableError`:
+
+```golang
+type ComposableError struct {
+	Error error
+	// This indicates that the consuming Reconcile function should return this error
+	ShouldBeReturned bool
+}
+```
+
+The type `ComposableError` contains the error, if any, and a boolean `ShouldBeReturned` to indicate whether the calling reconciler should return this error or not. This is used to distinguish errors that are minor and could be fixed by immediately retrying from errors that may indicate a stronger failure, such as an ill-formed yaml (in which case there is no need to retry right away). If there are no errors, then the `ResolveObject` returns nil.
```

Comment:
Given my previous comment, it seems like there could be a gray area for deciding how `ShouldBeReturned` should be set. I wonder if it would make sense to expose a separate helper function in the composable SDK to handle this such that `ResolveObject()` could return the error directly.

For example:
```go
func IsUnresolvableError(err error) bool { ... }
``` 

```go
err := composable.ResolveObject( ... )
if composable.IsUnresolvableError(err) {
    return reconcile.Result{}, err
} else if apierrors.IsNotFound(err) {
    // do some extra logic
} else {
    return reconcile.Result{}, err
}
```

This way callers could use composable's helper function OR define their own custom error handling, and it would get error handling logic out of `ResolveObject`.

Reply:
Yes it's a good idea to separate this concern, and add more detail to the kinds of error situations that can happen. Will open a work item.

Reply:
@joelanford The latest version of Composable SDK (v0.1.4) got rid of ComposableError. We now have a series of functions as you suggested:
```
func IsIllFormedRef(err error) bool 

func IsKindNotFound(err error) bool 

func IsObjectNotFound(err error) bool 

func IsValueNotFound(err error) bool 

func IsRefNotFound(err error) bool 
```
IsIllFormedRef means there is a problem with the way the cross-resource reference has been written in the yaml. IsKindNotFound means the kind for the reference does not exist. IsObjectNotFound means the object does not exist, and IsValueNotFound means the value does not exist within the object. The last one simply means either IsKindNotFound, or IsObjectNotFound, or IsValueNotFound is true.